vlasovsolver/gpu_pitch_angle_diffusion.cpp

```diff
@@ -0,0 +1,799 @@
+      for (size_t CellIdx = 0; CellIdx < numberOfLocalCells; CellIdx++) { // Iterate over all spatial cells
+         // Add elements to remapped cellIdx array
+         host_remappedCellIdxArray[CellIdx] = remappedCellIdx;
+
+         if(spatialLoopComplete[CellIdx]){
+            continue;
+         }
+
+         const auto CellID                  = LocalCells[CellIdx];
+         SpatialCell& cell                  = *mpiGrid[CellID];
+
+         vmesh::LocalID numberOfVelocityBlocks = cell.get_number_of_velocity_blocks(popID);
+
+         // Add elements to cellIdx arrays
+         host_cellIdxStartCutoff[remappedCellIdx] = totalNumberOfVelocityBlocks;
+         host_smallCellIdxArray[remappedCellIdx] = CellIdx;
+
+         totalNumberOfVelocityBlocks += numberOfVelocityBlocks;
+         remappedCellIdx++;
+      } // End spatial cell loop
+
+      std::vector<Real> host_Ddt (maxCellIndex);
+
+      // Create device arrays
+      size_t *dev_cellIdxArray, *dev_velocityIdxArray, *dev_smallCellIdxArray, *dev_remappedCellIdxArray, *dev_cellIdxStartCutoff;
+      Real *dev_Ddt, *dev_potentialDdtValues;
+      int *dev_cellIdxKeys;
+
+      // Allocate memory
+      CHK_ERR( gpuMalloc((void**)&dev_cellIdxArray, maxBlockIndex*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_velocityIdxArray, maxBlockIndex*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_cellIdxStartCutoff, maxCellIndex*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_smallCellIdxArray, maxCellIndex*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_remappedCellIdxArray, numberOfLocalCells*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_Ddt, maxCellIndex*sizeof(Real)) );
+      CHK_ERR( gpuMalloc((void**)&dev_potentialDdtValues, maxCellIndex*blocksPerVelocityCell*sizeof(Real)) );
+      CHK_ERR( gpuMalloc((void**)&dev_cellIdxKeys, maxCellIndex*blocksPerVelocityCell*sizeof(int)) );
+
+      // Copy data to device
+      CHK_ERR( gpuMemcpy(dev_cellIdxStartCutoff, host_cellIdxStartCutoff.data(), maxCellIndex*sizeof(size_t), gpuMemcpyHostToDevice) );
+      CHK_ERR( gpuMemcpy(dev_smallCellIdxArray, host_smallCellIdxArray.data(), maxCellIndex*sizeof(size_t), gpuMemcpyHostToDevice) );
+      CHK_ERR( gpuMemcpy(dev_remappedCellIdxArray, host_remappedCellIdxArray.data(), numberOfLocalCells*sizeof(size_t), gpuMemcpyHostToDevice) );
+
+      // Initialize with zero values
+      CHK_ERR( gpuMemset(dev_fmu, 0.0, numberOfLocalCells*nbins_v*nbins_mu*sizeof(Realf)) );
+      CHK_ERR( gpuMemset(dev_fcount, 0, numberOfLocalCells*nbins_v*nbins_mu*sizeof(int)) );
+
+      totalThreadsPerBlock = Hashinator::defaults::MAX_BLOCKSIZE/2; //Using Hashinator::defaults::MAX_BLOCKSIZE/2 = 512 blocks can lead to better streaming multiprocessor occupancy
+      int maxThreadIndex = totalNumberOfVelocityBlocks;
+      blocksPerGrid = (maxThreadIndex+totalThreadsPerBlock-1)/totalThreadsPerBlock;
+
+      // Find cell indices
```

Comment:
More verbose commenting please, explaining also *why* something is done. :) I think I was able to deduce what's going on here, but it'll be helpful in the future to have proper comments and descriptions (especially as this approach you have here is something which I haven't used anywhere else in the code)